feat: added extensions tab

Majortheus opened the pull request:
When I was first visiting the web site, I wanted to see what extensions @diego3g was using without to search manually or changing my "extensions.json", so again I tried to implement what i thot was the most interesting approach, which was to adapt the "extensions.json" to create a extensions tab similar to the code, I didn't go deeper on the clone as I thot some items whore unnecessary , to challenge myself and to achieve a result I was proud about. Comments and suggestions are very much welcome 😄, especially because there are some struct/folder changes to use and abuse the new layout systems of NextJS 13.

Preview:
![preview](https://user-images.githubusercontent.com/31166233/220619818-bed00429-678c-4227-b9eb-cfc238677e11.gif)


BTW, Catppuccin extensions seems like a cool theme, and Polacode a interesting extension, this was one of the reasons I did this PR, to have a peek on what could work for me...

Majortheus committed:
feat: added extensions tabs

Majortheus committed:
chore: adjusts to complete build and loading on extensions

b2evandro commented:
![image](https://user-images.githubusercontent.com/55710708/221444079-4a073312-ed63-45f5-8181-f80fb29956b1.png)

diego3g commented on src/app/extensions/[extension]/page.tsx:
```diff
@@ -0,0 +1,133 @@
+          {/* eslint-disable-next-line @next/next/no-img-element */}
+          <img
+            src={
+              extension.versions[0].files.find(
+                (file) =>
+                  file.assetType ===
+                  "Microsoft.VisualStudio.Services.Icons.Default"
+              )?.source ?? ""
```
This could be a variable above `return`.

diego3g commented on src/app/extensions/[extension]/page.tsx:
```diff
@@ -0,0 +1,133 @@
+          <div className="mt-2">
+            <Link
+              target="_blank"
+              href={`https://marketplace.visualstudio.com/items?itemName=${extension.publisher.publisherName}.${extension.extensionName}`}
```
As we are already displaying all the information about the extension inside this page, I think this link could redirect the user to the install itself, instead of redirecting to the marketplace URL.

diego3g commented on src/components/Extensions/ExtensionItem.tsx:
```diff
@@ -0,0 +1,42 @@
+          className="min-w-[42px] h-[42px]"
+        />
+      </div>
+      <div className="flex flex-col leading-5 text-sm text-[#908caa] max-w-full w-[calc(256px-42px-16px-24px)]"> {/* calc = 100% parent width (256) - img width (42) - parent padding left (16) - padding from scroll (24) */}
```
I don't think this calc is necessary while using flex or grid 🤔

Majortheus replied:
This is with the calc
![image](https://user-images.githubusercontent.com/31166233/224522020-093332c5-efa4-494f-aa43-48fcc3f80367.png)

This is without
![image](https://user-images.githubusercontent.com/31166233/224522039-8081f0fb-4636-4129-ac0e-21afc140511c.png)

basically the div element doesn't respect the parents width, and text oveflow "ellipsis" will not show because its hidden behind the content panel, if you have suggestions on how to make this div respect it's parent width, I will happily make the change, I also didn't like the calc as it's confusing, but it was the only way to mimic visual code style, and to be a little bit less confusing I added the comment

diego3g commented on src/components/Menu/MenuButton.tsx:
```diff
@@ -1,17 +1,44 @@
+      <Icon
+        strokeWidth={1.5}
+        size={28}
+        className="text-[#8F8CA8] data-[active=true]:text-[#E0DEF2]"
```
Did this work? I think you need to set a `group` class in parent `div` and use `group-data-[active=true]` here.

diego3g commented on src/app/extensions/[extension]/page.tsx:
```diff
@@ -0,0 +1,133 @@
+        </div>
+      </div>
+      <div className="max-h-full text-sm mt-8 extensions-markdown">
+        <ReactMarkdown rehypePlugins={[rehypeRaw]}>{detailsText}</ReactMarkdown>
```
What are `react-markdown` and `rehype` responsibilities here? I mean, what each one is doing?

Majortheus replied:
The detailsText it's in markdown format by default, and I thot it was easy to have a lib to handle that, but when I was testing the results, I came across some extensions that have html inside it's markdowns, and rehybe plugins handle that rendering inside the markdown, otherwise the text was rendering <a> </a> as text and I couldn't use css styles to make exactly like visual studio

diego3g reviewed (changes requested):
First, great job!

1. By creating a new `explorer` folder inside `app`, the URL not have a `explorer` prefix on all pages. Maybe we could rename it to `(explorer)` so it will not affect the URL, what do you think?
2. Instead of the blank page in `explorer/page.tsx`, maybe we could have a `404.tsx`;

Majortheus commented:
OMG, just saw the newest Rocketseat video, completely forgot about this PR, sorry... it has been a really busy couple of weeks, I will take a look real soon

dilincoln commented:
Hi everyone, I was testing this beautiful feature and I faced this issue below:

![CleanShot 2023-03-10 at 14 43 48](https://user-images.githubusercontent.com/37885454/224386791-894eccd6-4681-4fe4-9bb8-fbb4089f6ee3.gif)

Majortheus committed:
fix: fixed some styles bugs, better named variables, install button redirects to visual code

Majortheus committed:
fix: added early return because build uses not valid id as param

Majortheus commented:
Changes made, I think it didn't change much, some changes was to make (explorer) work, because it was hard to think a way to make any URL work as active except the ones that are already in use in menu, but I found a way, not sure if it's the best one, but it was interesting to make work.

Did some refactoring as was pointed out, way better naming, simplified active icon on menu, as was not working correctly, fixed bug pointed by @dilincoln, not sure why was happening, but a new deploy seems to work.

It's still bothering me the fact that tabs is working in localhost, but not on deployed version, and I still have no clue on why that's happening.

P.S.: I was thinking that can be fun to make the others tabs work, like: 

- **Search:**  make a full search on site and shows a link with maybe #:~:text=<text> to highlight the text on web page, the git showing 
- **Git/Source Control:** Show what PRs was accepted on this repository, or what changes was made to file texts or what extension went from active to unactive, or vice versa, like a history log, for instance: on May 13 @diego3g changed his default icons to material icons, or something like that
- **Debug**: I don't have many ideas on what to do with this tab, I was thinking some along the lines of showing the "code" behind the page
- **Remote Explorer**: I think that would be awesome if @diego3g could make other Rocketseat educator like Mayk Brito and Rodrigo Gonçalves and any other Rocketseat  Educator that has shown their visual code to make their fork and fill with their information, and make this tab show their website as "remote workspaces" or something along this lines to show others "fala devs" 
- **Experiments**: I don't have this tab on my visual code, I don't know what it usually show, but maybe could be a suggestion box?

I think it would be incredible but extremely hard to pull off, and most of the ideas I don't know how exactly I would do it, so I would love to hear what you guys think, and if it is realistic to expend that much time on this project that basically has done what was build for.
